### PR TITLE
InelasticScatteredTransmissionCorrectionMD algorithm

### PR DESCRIPTION
**Pull request by jmborr**
Companion to PR #31743

**Description of issue**  
See [SPEC106](https://code.ornl.gov/sns-hfir-scse/spectroscopy/spectroscopy/-/issues/106) and  [SPEC95](https://code.ornl.gov/sns-hfir-scse/spectroscopy/spectroscopy/-/issues/95)

**Description of work.**
- [x] Algorithm source
- [x] Unit tests
- [x] Workbench test
- [x] Release notes
- [x] documentation

**To test:**  
Run the following script on Mantid's workbench
```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

we1 = CreateSampleWorkspace(InstrumentName='HYSPEC',
                            WorkspaceType='Event',
                            Function='Flat background',
                            BankPixelWidth=1,
                            XUnit='DeltaE',
                            XMin=-10, 
                            XMax=19, 
                            BinWidth=0.1)
AddSampleLog(Workspace=we1,LogName='deltaE-mode', LogText='Direct', LogType='String')
AddSampleLog(Workspace=we1,LogName='Ei', LogText='20.', LogType='Number')
MoveInstrumentComponent(Workspace=we1, ComponentName='bank1', X=3, Z=3, RelativePosition=False)
MoveInstrumentComponent(Workspace=we1, ComponentName='bank2', X=-3, Z=-3, RelativePosition=False)

we2=CloneWorkspace(we1)

SetGoniometer(Workspace=we1, Axis0='0,0,1,0,1')
SetGoniometer(Workspace=we2, Axis0='30,0,1,0,1')

# new way
md1 = ConvertToMD(InputWorkspace=we1, QDimensions='Q3D')
md2 = ConvertToMD(InputWorkspace=we2, QDimensions='Q3D')
md = MergeMD(InputWorkspaces='md1,md2')

c = 1./11
mdstcMD = InelasticScatteredTransmissionCorrectionMD(InputWorkspace=md, ExponentFactor=c)


# old way
we1s = ScaleX(InputWorkspace=we1, Factor=-20., Operation='Add') # deltaE - Ei becomes -Ef
we1s = ScaleX(InputWorkspace=we1s, Factor=-1.0, Operation='Multiply') # the X-axis becomes Ef
# The ExponentialCorrection algorithm multiplies signal by C0*exp(-C1*x), where x is Ef
we1s = ExponentialCorrection(Operation="Multiply", InputWorkspace=we1s, C0=1., C1=-c) 
we1s = ScaleX(InputWorkspace=we1s, Factor=-1.0, Operation='Multiply') # the X-axis becomes -Ef
we1s = ScaleX(InputWorkspace=we1s, Factor=20., Operation='Add') # Ei - Ef converts back to DeltaE
md1s = ConvertToMD(InputWorkspace=we1s, QDimensions='Q3D')

we2s = ScaleX(InputWorkspace=we2, Factor=-20., Operation='Add')
we2s = ScaleX(InputWorkspace=we2s, Factor=-1.0, Operation='Multiply')
we2s = ExponentialCorrection(Operation="Multiply",InputWorkspace=we2s,C0=1.,C1=-c)
we2s = ScaleX(InputWorkspace=we2s, Factor=-1.0, Operation='Multiply')
we2s = ScaleX(InputWorkspace=we2s, Factor=20., Operation='Add')
md2s = ConvertToMD(InputWorkspace=we2s, QDimensions='Q3D')
mds = MergeMD(InputWorkspaces='md1s, md2s')

CompareMDWorkspaces(Workspace1='mdstcMD', Workspace2='mds', Tolerance=0.0001, CheckEvents=False)

# Integrate over all  Q-coordinates, then bin over the remaining `DeltaE` coordinate.
# Histograms for `mds` and`mdstcMD` should match.
mdsHisto = BinMD(mds, AxisAligned=True, AlignedDim0="Q_sample_x, -3,3,1", AlignedDim1="Q_sample_y, -3,3,1", 
                 AlignedDim2="Q_sample_z, -1,7,1", AlignedDim3="DeltaE, -10,19,100")
mdstcMDHisto = BinMD(mdstcMD, AxisAligned=True, AlignedDim0="Q_sample_x, -3,3,1",
                     AlignedDim1="Q_sample_y, -3,3,1", AlignedDim2="Q_sample_z, -1,7,1",
                     AlignedDim3="DeltaE, -10,19,100")

# Plot the histograms in  `DeltaE`. They should overlap
fig, ax = plt.subplots(subplot_kw={'projection':'mantid'})
ax.errorbar(mdsHisto,'rs')
ax.errorbar(mdstcMDHisto,'bo')
fig.show()
```

*There is no associated issue.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/MDAlgorithms/src/HYSPECScatteredTransmissionCorrectionMD.cpp [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+
+std::map<std::string, std::string> HYSPECScatteredTransmissionCorrectionMD::validateInputs() {
+  std::map<std::string, std::string> output;
+  // validate input workspace
+  std::string workspace_error = checkInputWorkspace();
+  if (workspace_error != "")
+    output["InputWorkspace"] = workspace_error;
+  // ExponentFactor should be positive
+  double c = getProperty("ExponentFactor");
+  if (c <= 0.0)
```
Comment (AndreiSavici) [inline Framework/MDAlgorithms/src/HYSPECScatteredTransmissionCorrectionMD.cpp:62]: Can't get c<0 due to validator. If you make it exclusive in the `init`, you don't need to check it at all

File: Framework/MDAlgorithms/src/HYSPECScatteredTransmissionCorrectionMD.cpp [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+DECLARE_ALGORITHM(HYSPECScatteredTransmissionCorrectionMD)
+
+//---------------------------------------------------------------------------------------------------------
+
+void HYSPECScatteredTransmissionCorrectionMD::init() {
+  declareProperty(std::make_unique<WorkspaceProperty<IMDEventWorkspace>>("InputWorkspace", "", Direction::Input),
+                  "Input MDEventWorkspace. Either QSample (or QLab) frame plus DeltaE, or just |Q| plus DeltaE");
+
+  auto mustBePositive = std::make_shared<BoundedValidator<double>>();
+  mustBePositive->setLower(0.0);
```
Comment (AndreiSavici) [inline Framework/MDAlgorithms/src/HYSPECScatteredTransmissionCorrectionMD.cpp:43]: https://github.com/mantidproject/mantid/blob/master/Framework/Kernel/inc/MantidKernel/BoundedValidator.h#L38 Make it exclusive

File: Framework/MDAlgorithms/src/HYSPECScatteredTransmissionCorrectionMD.cpp [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    output["ExponentFactor"] = "ExponentFactor should be positive";
+  return output;
+}
+
+//---------------------------------------------------------------------------------------------------------
+
+std::string HYSPECScatteredTransmissionCorrectionMD::checkInputWorkspace() {
+  // Verify the input workspace if of type MDEventWorkspace
+  IMDEventWorkspace_sptr inputws = getProperty("InputWorkspace");
+  if (!inputws)
```
Comment (AndreiSavici) [inline Framework/MDAlgorithms/src/HYSPECScatteredTransmissionCorrectionMD.cpp:72]: This will never run, since it will be checked in the init

File: Framework/MDAlgorithms/src/HYSPECScatteredTransmissionCorrectionMD.cpp [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    return "Input workspace is not of type MDEventWorkspace";
+
+  // Verify the dimension of the input workspace
+  std::string dimensionError = checkInputMDDimensions();
+  if (dimensionError != "")
+    return dimensionError;
+
+  uint16_t nRuns(inputws->getNumExperimentInfo());
```
Comment (AndreiSavici) [inline Framework/MDAlgorithms/src/HYSPECScatteredTransmissionCorrectionMD.cpp:80]: This is not HYSPEC only.

File: docs/source/algorithms/HYSPECScatteredTransmissionCorrectionMD-v1.rst [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+.. summary::
+
+.. relatedalgorithms::
+
+.. properties::
+
+Description
+-----------
+
+The intensity :math:`I` of every event collected is corrected by:
```
Comment (AndreiSavici) [inline docs/source/algorithms/HYSPECScatteredTransmissionCorrectionMD-v1.rst:12]: Please replace the first sentence by "In instruments where there is a polarization analysis stage in the scattered beam, neutron intensity is reduced by an absorption factor. This has an exponential dependence on the scattered energy. To correct for this, the detected intensity :math:`I` of every event collected is corrected by:"

**Review by AndreiSavici — COMMENTED** [review]
Minor changes. Also, remove HYSPEC from the name of the algorithm (might be used at LET as well)

**Review by AndreiSavici — DISMISSED** [review]

**Review by AndreiSavici — DISMISSED** [review]

File: Framework/Algorithms/src/CreateSampleWorkspace.cpp [hunk 6]
```diff
@@ -128,6 +128,7 @@ void CreateSampleWorkspace::init() {
                   "Minimum (default) is 1 scan point, which gives a "
                   "non-scanning workspace.");
 
+  declareProperty("InstrumentName", "basic_rect", Direction::Input);
```
Comment (rosswhitfield) [inline Framework/Algorithms/src/CreateSampleWorkspace.cpp:131]: Why are you adding a new property to `CreateSampleWorkspace`? This seems unrelated to the rest of the work.

Comment (AndreiSavici) [reply Framework/Algorithms/src/CreateSampleWorkspace.cpp:131]: @rosswhitfield When Jose initially looked at the requirements, he thought that it only applies to HYSPEC. So he wanted to add an instrument name validation step. This was supposed to be the helper to write the test. The algorithm is more general than that, so the name validator is not required. But it might be useful for someone else.

Comment (rosswhitfield) [reply Framework/Algorithms/src/CreateSampleWorkspace.cpp:131]: OK but this new functionality should then have a unit test and its own release note

Comment (AndreiSavici) [reply Framework/Algorithms/src/CreateSampleWorkspace.cpp:131]: @jmborr ?

Comment (jmborr) [reply Framework/Algorithms/src/CreateSampleWorkspace.cpp:131]: I'll get to it

Comment (jmborr) [reply Framework/Algorithms/src/CreateSampleWorkspace.cpp:131]: @AndreiSavici I included Ross' requests. Can you reapprove, please?

**Review by AndreiSavici — APPROVED** [review]

**Comment by jmborr** [comment]
@AndreiSavici @rosswhitfield Thank you! 🕺
